contrib/docker/Dockerfile

```diff
@@ -12,6 +15,8 @@
 FROM golang@sha256:2381c1e5f8350a901597d633b2e517775eeac7a6682be39225a93b22cfd0f8bb AS builder
 RUN apk add --no-cache git ca-certificates upx
 
+ARG BUILD_TAG
```

Comment:
Is this supposed to be here a second time?  It doesn't look like it to me.

Reply:
Yes, this is the way a default value can be defined (first `ARG` before the `FROM`).  The second one imports the value into the build stage.

https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

Reply:
Thanks for the link.  I looked into it a bit and did some testing and you only need to specify the `ARG` before the `FROM` if you're going to use the argument in the `FROM` statement itself.  So, if we wanted an argument for the build version of Go, then it would need to be at the top.

However, since the build tag isn't needed in the `FROM`, I think it should to be just before it's used to avoid invalidating the cache of all layers that come before it since they don't change.

Also, I noticed this is only dealing with `dcrd` and not `dcrctl`, but they both have to be in sync.  It is also feasible that they don't use the same tags.

Given all that, I would suggest the following diff (and then update the `README.md` accordingly):

```diff
diff --git a/contrib/docker/Dockerfile b/contrib/docker/Dockerfile
index 28dd4271..0f06c287 100644
--- a/contrib/docker/Dockerfile
+++ b/contrib/docker/Dockerfile
@@ -32,15 +32,17 @@ RUN adduser \

 # Build dcrd and other commands it provides.
 WORKDIR /go/src/github.com/decred/dcrd
-RUN git clone https://github.com/decred/dcrd . && \
+ARG DCRD_BUILD_TAG=master
+RUN git clone --branch ${DCRD_BUILD_TAG} https://github.com/decred/dcrd . && \
     CGO_ENABLED=0 GOOS=linux \
     go install -trimpath -tags safe,netgo,timetzdata \

 # Build dcrctl.
 WORKDIR /go/src/github.com/decred/dcrctl
-RUN git clone https://github.com/decred/dcrctl . && \
+ARG DCRCTL_BUILD_TAG=${DCRD_BUILD_TAG}
+RUN git clone --branch ${DCRCTL_BUILD_TAG} https://github.com/decred/dcrctl . && \
     CGO_ENABLED=0 GOOS=linux \
     go install -trimpath -tags safe,netgo -ldflags="-s -w"
```

That makes it so that:

1) It reuses the cache for all layers that come before it when you change the build tag:
  ```
  Sending build context to Docker daemon  66.62MB
  Step 1/28 : FROM golang@sha256:2381c1e5f8350a901597d633b2e517775eeac7a6682be39225a93b22cfd0f8bb AS builder
   ---> feb4bbda921c
  Step 2/28 : RUN apk add --no-cache git ca-certificates upx
   ---> Using cache
   ---> ea6f84010571
  Step 3/28 : RUN mkdir /emptydatadir
   ---> Using cache
   ---> 9934dba39913
  Step 4/28 : ENV USER=decred
   ---> Using cache
   ---> 39175e261011
  Step 5/28 : ENV UID=10000
   ---> Using cache
   ---> c79d8c12e4b0
  Step 6/28 : RUN adduser     --disabled-password     --gecos ""     --home="/home/${USER}"     --shell "/sbin/nologin"     --no-create-home     --uid "${UID}"     "${USER}"
   ---> Using cache
   ---> bf37cef21ef3
  Step 7/28 : WORKDIR /go/src/github.com/decred/dcrd
   ---> Using cache
   ---> 8901e76d6ce0
  Step 8/28 : ARG DCRD_BUILD_TAG=master
   ---> Using cache
   ---> ab34862d381a
  Step 9/28 : RUN git clone --branch ${DCRD_BUILD_TAG} https://github.com/decred/dcrd . &&     CGO_ENABLED=0 GOOS=linux     go install -trimpath -tags safe,netgo,timetzdata       -ldflags="-s -w"       . ./cmd/gencerts
   ---> Running in 4d435395381b
```
2) It allows an independent tag/commitsh for `dcrd` and `dcrctl` with clearly identified names
3) It uses the `DCRD_BUILD_TAG` argument for `dcrctl` by default so only specifying that one will make `dcrctl` match

 
What do you think?

Reply:
Thank you for the patch, I've updated the PR.  Also disabled the git detached head warning.